### ISSUE TITLE
Fused AdamW causes NaN loss

### ISSUE THREAD

**Issue by oddlama** [pytorch/pytorch#95781]
### 🐛 Describe the bug

There already has been an extended discussion of this issue over on the nanoGPT repository:
- karpathy/nanoGPT#167

I have been encouraged to submit a separate bug report here, since the issue seems to lie with pytorch. I have prepared a preconfigured [fork of nanoGPT](https://github.com/oddlama/nanoGPT_nan) where I provided my configuration and part of my dataset which causes the issue to appear. It now immediately produces NaNs after the first training step, but only if fused adam is in use.

The settings are basically the nanoGPT shakespeare configuration, but using a blocksize of 343 and vocab size of 2006. The data is quite sparse (only 1,3% is not 0, which significantly accelerates getting to the issue). I've included 1000 batches of my own real data so you will have the same conditions as I do. I'm training on a single 2080Ti, please look at a diff of the latest commit in this repository to see what I changed exactly - it's not much.

```bash
> git clone https://github.com/oddlama/nanoGPT_nan
> cd nanoGPT_nan
```
```bash
> python train.py config/nan.py --allow_fused=True
[...]
number of parameters: 11.39M
using fused AdamW: True
compiling the model... (takes a ~minute)
step 0: train loss 6.6010, val loss 6.6009
[2023-03-01 15:19:43,485] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
iter 0: loss 6.6793, time 22316.39ms, mfu -100.00%
iter 1: loss nan, time 3249.91ms, mfu -100.00%
```
```bash
> python train.py config/nan.py --allow_fused=False
[...]
number of parameters: 11.39M
using fused AdamW: False
compiling the model... (takes a ~minute)
step 0: train loss 6.6010, val loss 6.6009
[2023-03-01 15:20:26,110] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/nanoGPT_nan/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
iter 0: loss 6.6793, time 22286.14ms, mfu -100.00%
iter 1: loss 6.6811, time 3534.44ms, mfu -100.00%
iter 2: loss 6.6770, time 3365.06ms, mfu -100.00%
iter 3: loss 4.7208, time 3366.52ms, mfu -100.00%
iter 4: loss 3.0927, time 3365.22ms, mfu -100.00%
iter 5: loss 2.0694, time 3362.80ms, mfu 6.51%
iter 6: loss 1.4752, time 3360.81ms, mfu 6.51%
# continues to work fine
```

Hope this helps.

P.S.:

Can anyone explain what the warning means? Doesn't seem to be related to the issue, but is it a user error that should be fixed?
```
[2023-03-01 15:19:43,485] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0.dev20230220+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.25.0
Libc version: glibc-2.37

Python version: 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.15.79.1-microsoft-standard-WSL2-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 516.94
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen 9 5900X 12-Core Processor
CPU family:                      25
Model:                           33
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
Stepping:                        0
BogoMIPS:                        7400.01
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr arat umip vaes vpclmulqdq rdpid fsrm
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       384 KiB (12 instances)
L1i cache:                       384 KiB (12 instances)
L2 cache:                        6 MiB (12 instances)
L3 cache:                        32 MiB (1 instance)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.0.0+d54c04abe2
[pip3] torch==2.0.0.dev20230220+cu118
[pip3] torchaudio==2.0.0.dev20230223+cu118
[pip3] torchinfo==1.7.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.15.0.dev20230223+cu118
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD @janeyx99

**Comment by janeyx99** [comment]
cc @crcrpar This is pretty concerning--do you have any thoughts? Also, we most likely want to add AMP x fused adam(w) unit tests as it looks to be a common source of wrongness.

**Comment by crcrpar** [comment]
I'll look into it for sure.
though just so you know, there's a case of fused ones with grad scalers: https://github.com/pytorch/pytorch/blob/e5a959a2d49e627905c2ad58314a85c9e2f2c1df/test/test_cuda.py#L2462

**Comment by bdhirsh** [comment]
> Can anyone explain what the warning means? Doesn't seem to be related to the issue, but is it a user error that should be fixed?

```
[2023-03-01 15:19:43,485] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
```

@oddlama nope - the first warnings is just letting you know that the results of operators involving randomness won't give identical results across eager mode and `torch.compile`. The second warning is a bug that you can ignore - it was recently fixed at https://github.com/pytorch/pytorch/pull/95521

**Comment by atalman** [comment]
cc @crcrpar  @janeyx99 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176

**Comment by crcrpar** [comment]
Hi @atalman,
thank you for the information, I wrote down the PR for this issue in https://github.com/pytorch/pytorch/issues/97272#issuecomment-1488878531 :)
